### ISSUE TITLE
[Epic] SignalR Transport Updates

### ISSUE THREAD

**Issue by anurse** [dotnet/aspnetcore#21104]
The idea here is to build a new transport based on HTTP streaming. The idea is for the client to establish **two** long-running HTTP requests. Termination of either request by either party tears down the whole connection (and terminates the other request). This transport is designed for HTTP/2 environments, and the transport would likely be designed to function only when the requests are actually using HTTP/2.

* One request serves as the "upstream" connection. The client sends the request headers (including auth) and then streams data to the server. The server **does not write any response** to this request until the connection terminates.

* The other request serves as the "downstream" connection. The client sends request headers and an empty body. The server sends a long-running response back and writes data to it as data becomes available. The server concludes the response when the connection is terminated.

**Why two requests?** We could have a single request with both the client and server streaming request/response data. This is permitted in the HTTP protocol. However, support for this in all platforms (Browser, .NET, Node, Java, etc.) is limited and inconsistent. Doing this two-stream model **doesn't** stop us from doing a bi-directional transport in the future.

With this transport, we can also consider deprecating Server-Sent Events. The SSE transport provides a middle-ground between Long Polling and WebSockets and has served us well but it has limitations:

* IE still doesn't support it, so it's usefulness as a "downlevel" transport is limited. The set of browsers that support WebSockets is a superset of the set of browsers that support SSE
* It can be impacted by proxies, though not as badly as WebSockets. We've encountered proxies that buffer responses which delays messages and can break connections (if timeouts elapse)
* It is limited to the JSON protocol. SSE is an inherently text-based protocol. It is sensitive to line-breaks, making it difficult to encode binary messages. While Base64 encoding is an option, we haven't use that and simply avoid using the SSE transport for MessagePack.

----

Work Items:
* [ ] Update the [Transport Protocols](https://github.com/dotnet/aspnetcore/blob/master/src/SignalR/docs/specs/TransportProtocols.md) spec to include the new HTTP Streaming transport
* [ ] Implement the HTTP Streaming transport in the .NET Server and Client
* [ ] Implement the HTTP Streaming transport in the JavaScript Client
* [ ] Update documentation on [SignalR Client Features](https://docs.microsoft.com/en-us/aspnet/core/signalr/client-features?view=aspnetcore-3.1) for 5.0
* [ ] https://github.com/dotnet/aspnetcore/issues/20462 Remove ServerSentEvents transport

**Comment by BrennanConroy** [comment]
Should do something about https://github.com/dotnet/aspnetcore/issues/12662

**Comment by davidfowl** [comment]
OR close this as a dupe 😄

**Comment by BrennanConroy** [comment]
This is the epic, so we'll keep it open

**Comment by anurse** [comment]
Yep, was about to close that actually.

**Comment by davidfowl** [comment]
Also we're learning there are big issues around how clients behave when they hit various limits in HTTP/2 especially because we plan on having long running requests.

TL;DR Some clients don't open new connections when they hit the concurrent streams limit.

**Comment by halter73** [comment]
> TL;DR Some clients don't open new connections when they hit the concurrent streams limit.

Context: https://github.com/dotnet/runtime/issues/35088

Concurrent connection limits have always been an issue with SignalR. Browsers and the .NET framework used to enforce really low per-host concurrent connection limits. Browsers still do enforce low limits for non-WebSocket connections.

Over time, browsers have gradually increased their WebSocket limits. Now it seems to be [up to 255](https://chromium.googlesource.com/chromium/src/net/+/master/socket/client_socket_pool_manager.cc#48) concurrent WebSocket connections per host. Firefox now defaults to 200 if you look at network.websocket.max-connections in about:config.

If a server uses 100 SETTINGS_MAX_CONCURRENT_STREAMS like Kestrel does by default, that's at most 50 concurrent SignalR "connections" using this new streaming transport over HTTP/2. That by itself doesn't seem so bad, but unlike with WebSockets (which have their own connection pool), hitting this limit could affect "normal" requests.

We'll probably want to use negotiate or something to detect HTTP/2, I doubt we'd want to do this using HTTP/1.1 requests.
